File: src/client/java/fr/hardel/asset_editor/client/javafx/components/ui/tree/TreeUtils.java
package fr.hardel.asset_editor.client.javafx.components.ui.tree;

public final class TreeUtils {

    public static boolean hasActiveDescendant(TreeNodeModel node, String activeId) {
        if (activeId == null || activeId.isBlank()) return false;
        if (activeId.equals(node.elementId())) return true;
        for (TreeNodeModel child : node.children().values()) {
            if (hasActiveDescendant(child, activeId)) return true;
        }
        return false;
    }

    public static int calculateCount(TreeNodeModel node) {
        if (node.elementId() != null && !node.elementId().isBlank()) return 1;
        int count = 0;
        for (TreeNodeModel child : node.children().values()) {
            count += calculateCount(child);
        }
        node.setCount(count);
        return count;
    }

    private TreeUtils() {
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/ui/tree/TreeUtils.kt
package fr.hardel.asset_editor.client.compose.components.ui.tree

object TreeUtils {

    fun hasActiveDescendant(node: TreeNodeModel, activeId: String?): Boolean {
        if (activeId.isNullOrBlank()) return false
        if (activeId == node.elementId) return true
        return node.children.values.any { hasActiveDescendant(it, activeId) }
    }

    fun recalculateCount(node: TreeNodeModel): Int {
        if (!node.elementId.isNullOrBlank()) return 1
        val total = node.children.values.sumOf { recalculateCount(it) }
        node.count = total
        return total
    }
}
